eliot-service/eliot/downloader.py

```diff
@@ -0,0 +1,157 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, you can obtain one at http://mozilla.org/MPL/2.0/.
+
+"""
+Contains code for downloading sym files from one or more sources.
+"""
+
+import markus
+
+from eliot.librequests import session_with_retries
+from eliot.libsentry import get_sentry_client
+
+
+METRICS = markus.get_metrics(__name__)
+
+
+class FileNotFound(Exception):
+    """File was not found because it doesn't exist."""
+
+
+class ErrorFileNotFound(Exception):
+    """File was not found because of possible transient error."""
+
+
+class Source:
+    """Defines a source manager
+
+    This takes a source url and then uses that to retrieve SYM files from the source.
+
+    """
+
+    def __init__(self, source_url):
+        self.source_url = source_url
+
+    def get(self, debug_filename, debug_id, filename):
+        """Retrieve a source url.
+
+        :arg str debug_filename: the debug_filename
+        :arg str debug_id: the debug_id
+        :arg str filename: the symbol filename
+
+        :returns: bytes
+
+        :raises FileNotFound: if the file cannot be found
+
+        :raises ErrorFileNotFound: if the file cannot be found because of some possibly
+            transient error like a timeout or a connection error
+
+        """
+        raise NotImplementedError
+
+
+class HTTPSource(Source):
+    """Source for HTTP/HTTPS requests."""
+
+    def __init__(self, source_url):
+        self.source_url = source_url.rstrip("/") + "/"
+        self.session = session_with_retries()
+
+    def _make_key(self, debug_filename, debug_id, filename):
+        """Generates a key from given arguments
+
+        :arg str debug_filename: the debug_filename
+        :arg str debug_id: the debug_id
+        :arg str filename: the symbol filename
+
+        :returns: key as a str
+
+        """
+        # NOTE(willkg): This has to match Tecken's upload code
+        return "%s/%s/%s" % (debug_filename, debug_id, filename)
+
+    def get(self, debug_filename, debug_id, filename):
+        """Retrieve a source url.
+
+        :arg str debug_filename: the debug_filename
+        :arg str debug_id: the debug_id
+        :arg str filename: the symbol filename
+
+        :returns: bytes
+
+        :raises FileNotFound: if the file cannot be found
+
+        :raises ErrorFileNotFound: if the file cannot be found because of some possibly
+            transient error like a timeout or a connection error
+
+        """
+        key = self._make_key(debug_filename, debug_id, filename)
+        url = "%s%s" % (self.source_url, key)
+
+        resp = self.session.get(url, allow_redirects=True)
+        if resp.status_code != 200:
+            METRICS.incr("sym_download", tags=["statuscode:%d" % resp.status_code])
+
+            # If the status_code is 404, that's a legitimate FileNotFound. Anything else
+            # is either fishy or a server error.
```

Comment:
Note that the session is already set up to retry a whole series of conditions, so this is more like "anything else at this point after retrying a bunch of ephemeral universe hooey".